integration_tests/src/main/python/arithmetic_ops_test.py

```diff
@@ -973,35 +977,52 @@ def test_greatest(data_gen):
                 f.greatest(*command_args)))
 
 
-def _test_div_by_zero(ansi_mode, expr):
-    ansi_conf = {'spark.sql.ansi.enabled': ansi_mode == 'ansi'}
+def _test_div_by_zero(ansi_mode, expr, exception):
+    ansi_conf = {'spark.sql.ansi.enabled': ansi_mode}
     data_gen = lambda spark: two_col_df(spark, IntegerGen(), IntegerGen(min_val=0, max_val=0), length=1)
     div_by_zero_func = lambda spark: data_gen(spark).selectExpr(expr)
-    if is_before_spark_320():
-        err_message = 'java.lang.ArithmeticException: divide by zero'
-    elif is_before_spark_330():
-        err_message = 'SparkArithmeticException: divide by zero'
-    elif is_before_spark_340() and not is_databricks113_or_later():
-        err_message = 'SparkArithmeticException: Division by zero'
-    else:
-        err_message = 'SparkArithmeticException: [DIVIDE_BY_ZERO] Division by zero'
 
-    if ansi_mode == 'ansi':
+    if ansi_mode:
         assert_gpu_and_cpu_error(df_fun=lambda spark: div_by_zero_func(spark).collect(),
                                  conf=ansi_conf,
-                                 error_message=err_message)
+                                 error_message=exception)
     else:
         assert_gpu_and_cpu_are_equal_collect(div_by_zero_func, ansi_conf)
 
-
-@pytest.mark.parametrize('expr', ['1/0', 'a/0', 'a/b'])
-def test_div_by_zero_ansi(expr):
-    _test_div_by_zero(ansi_mode='ansi', expr=expr)
-
-@pytest.mark.parametrize('expr', ['1/0', 'a/0', 'a/b'])
-def test_div_by_zero_nonansi(expr):
-    _test_div_by_zero(ansi_mode='nonAnsi', expr=expr)
-
+div_by_zero_data=['1/0', 'a/0', 'a/b']
+@pytest.mark.parametrize('expr, exception',
+    # is_before_320
+    [pytest.param(i, 'java.lang.ArithmeticException: divide by zero',
+                  marks=pytest.mark.skipif(not is_before_spark_320(),
+                                           reason="Skipping for Spark version >= 3.2.0"))
+                    for i in div_by_zero_data]
+    +
+    # else if is_before_330
+    [pytest.param(i, 'SparkArithmeticException: divide by zero',
+                 marks=pytest.mark.skipif(not (not is_before_spark_320() and is_before_spark_330()),
+                                          reason="Skipping for 3.2.0 > Spark version >= 3.3.0"))
+                    for i in div_by_zero_data]
+    +
+    # else if is_before_340 and not is_databricks_113_or_later
+    [pytest.param(i, 'SparkArithmeticException: Division by zero',
+                 marks=pytest.mark.skipif(not (not is_before_spark_330() and is_before_spark_340() and not is_databricks113_or_later()),
+                                          reason="Skipping for 3.3.0 > Spark version >= 3.4.0 or db >= 11.3"))
+                 for i in div_by_zero_data]
```

Comment:
Curious why parameterizing on the same data values based on expected exception rather than have the test directly compute the expected exception type based on spark version like is done above for test_cast_neg_to_decimal_err?  It's much harder to read this way and figure out what is being tested and what isn't.

Reply:
I am not very happy with this PR but I wanted to know what others think. If there was a way to know if an expression was a `literal`, this would be a smaller change, but I don't think there is a way in pyspark. 

The key difference in this case and the ones previously is that Spark 3.4 is deciding to throw this exception only when the expression is a `literal`, but throws a different one when it's not a `literal`

Reply:
Then split it into two tests, one that tests literals and one that doesn't. You can factor out the common code between the two tests as utility methods.

Reply:
Ironically that was my very first thought but I talked myself out of it because there are potentially more changes like this one coming up and I wanted to have a generic solution, granted that this PR is also not a generic solution. 

I will update

Reply:
yes, this turned very different from what I expected when we ended pair programming. We should not pursue it in this form